ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,445 @@
+                                        0.0f, std::min<float>(x1, img_W - coordinates_offset));
+                                    y1 = std::max<float>(
+                                        0.0f, std::min<float>(y1, img_H - coordinates_offset));
+                                }
+
+                                // recompute new width & height
+                                const float box_w = x1 - x0 + coordinates_offset;
+                                const float box_h = y1 - y0 + coordinates_offset;
+
+                                proposals[proposal_off + anchor].x0 = static_cast<T>(x0);
+                                proposals[proposal_off + anchor].y0 = static_cast<T>(y0);
+                                proposals[proposal_off + anchor].x1 = static_cast<T>(x1);
+                                proposals[proposal_off + anchor].y1 = static_cast<T>(y1);
+                                proposals[proposal_off + anchor].score =
+                                    static_cast<T>((min_box_W <= box_w) * (min_box_H <= box_h)) *
+                                    score;
+                            }
+                        }
+                    }
+                }
+
+                template <typename T>
+                static void nms(const int num_boxes,
+                                std::vector<int>& is_dead,
+                                const std::vector<ProposalBox<T>>& proposals,
+                                std::vector<unsigned int>& index_out,
+                                int& num_out,
+                                const int base_index,
+                                const float nms_thresh,
+                                const int max_num_out,
+                                T coordinates_offset)
+                {
+                    std::fill(is_dead.begin(), is_dead.begin() + num_boxes, 0);
+                    for (int box = 0; box < num_boxes; ++box)
+                    {
+                        if (is_dead[box])
+                            continue;
+
+                        index_out[num_out++] = base_index + box;
+                        if (num_out == max_num_out)
+                            break;
+
+                        int tail = box + 1;
+                        for (; tail < num_boxes; ++tail)
+                        {
+                            float res = 0.0f;
+
+                            const T x0i = proposals[box].x0;
+                            const T y0i = proposals[box].y0;
+                            const T x1i = proposals[box].x1;
+                            const T y1i = proposals[box].y1;
+
+                            const T x0j = proposals[tail].x0;
+                            const T y0j = proposals[tail].y0;
+                            const T x1j = proposals[tail].x1;
+                            const T y1j = proposals[tail].y1;
+
+                            if (x0i <= x1j && y0i <= y1j && x0j <= x1i && y0j <= y1i)
```

Comment:
Consider to add some meaningful name for this check (if it exist in algo description of course).

Reply:
I mean more like this:
```
const bool first_box_begins_before_second_end = x0i <= x1j && y0i <= y1j ;
const bool second_box_begins_before_the_first_ends = x0j <= x1i && y0j <= y1i;
if ( first_box_begins_before_second_end  && second_box_begins_before_the_first_ends ){
//...
```